project-koku/koku
# Consolidate Sources APIs and enable PATCH

The sources `/billing_source` and `/authentication` were very basic.  Designed to provide just enough functionality to enable basic backend sources integration.

Now that we have integration well underway and I've gotten some feedback from the API consumers (UI and QE), I'd like to make some changes to simplify the Koku-Sources API interactions.

With this change there is now a single `/sources` endpoint (also a Koku-api proxy) which takes the `source_id` as a path parameter.  The GET response provides information into the source which was previously only accessible via direct db calls.  On the Sources Client there is an internal `/sources/<source_id>/status/` endpoint which returns the same thing as `/source-status`.  This internal API hook to be used by the platform to provide source status information.  Work here is still underway so this calling convention is subject to change.

Now API consumers will be able to use `PATCH` to set the authentication and billing_source information on a source. (yes, for Azure you can set both with a single API call!)

I'm keeping the old `/billing_source` and `/authentication` endpoints around until we move QE tests and UI code to the new `/sources` API.

The API also respects the user's account ID and only returns sources that belong to the users account.

**Testing**
1. Add AWS Source to platform with auth_header account id 10001.
2.  Attempt to add bucket with auth_header account id 10002.  Verify that the /sources API returns 404.
3.  Access internal status api for source id 1 and verify source is unavaiable.
4.  Add bucket with auth_header account id 10001.  Verify provider is created successfully

**Test Results**
[sources_api_enhance_ut.txt](https://github.com/project-koku/koku/files/3918929/sources_api_enhance_ut.txt)

Addresses #1439 

cc: @rvsia @boaz0
